Design/indexing_rollout_notes.md

```diff
@@ -0,0 +1,209 @@
+  bool isMonitor(const size_t index) const;
+  double l2(const size_t index) const;
+  Detector_sptr uniqueDetector(const size_t index) const; // Throws if no or more than 1 detector
+  bool hasUniqueDetector(const size_t index) const; // Should this be part of IndexInfo instead?
+};
+```
+
+There is also another (special?) case (example from current code):
+
+```cpp
+Geometry::IDetector_const_sptr det;
+size_t numDetectors = workspace->getSpectrum(wi).getDetectorIDs().size();
+if (numDetectors > 1) {
+  Instrument_const_sptr inst = workspace->getInstrument();
+  det = inst->getDetector(*workspace->getSpectrum(wi).getDetectorIDs().begin());
+} else {
+  det = workspace->getDetector(wi);
+}
+```
+
+## Duplicate spectrum numbers
+
+Currently quite a few algorithms or tests use or create workspaces with non-unique spectrum numbers.
+Ultimately we want to forbid this, but it is not yet clear if we should attempt to clean this up right away, or wait until `IndexInfo` has developed into a more stable state.
+
+
+## Standards
+
+#### Spectrum numbers starting at 0
+
+Currently quite a few algorithms that create workspaces where spectrum numbers do not start at 1 but at 0.
+This is non standard.
+Should this be cleaned up as part of this refactoring effort?
+
+Examples:
+
+- `AppendSpectra`
+
+
+## MatrixWorkspace
+
+There is a considerable number of methods in `MatrixWorkspace` that could potentially be replaced or removed during this refactoring effort.
+The detailed strategy should depend on how and where these methods are used.
+
+```cpp
+class MatrixWorkspace {
+public:
+  void updateSpectraUsing(const SpectrumDetectorMapping &map);
+  void rebuildSpectraMapping(const bool includeMonitors);
+  spec2index_map getSpectrumToWorkspaceIndexMap() const;
+  detid2index_map getDetectorIDToWorkspaceIndexMap(bool throwIfMultipleDets) const;
+  std::vector<size_t> getDetectorIDToWorkspaceIndexVector(detid_t &offset, bool throwIfMultipleDets) const;
+  std::vector<size_t> getSpectrumToWorkspaceIndexVector(specnum_t &offset) const;
+  std::vector<size_t> getIndicesFromSpectra(const std::vector<specnum_t> &spectraList) const;
+  size_t getIndexFromSpectrumNumber(const specnum_t specNo) const;
+  std::vector<size_t> getIndicesFromDetectorIDs(const std::vector<detid_t> &detIdList) const;
+  std::vector<specnum_t> getSpectraFromDetectorIDs(const std::vector<detid_t> &detIdList) const;
+};
+```
+
+
+## Notes about individual algorithms
+
+The following is a more or less random selection/collection about specific algorithms or classes and how the could or should change.
+
+#### SumSpectra
+
+Uses smallest spectrum number in sum as spectrum number for output.
+This does not seem to make much sense, maybe it is just an arbitrary requirement by an individual piece of client code.
+It would probably be easier and cleaner to simply remove special handling of spectrum numbers from `SumSpectra` and adapt the client code in question.
```

Comment:
Actually it seems pretty simple. It makes the changes in `SumSpectra` itself trivial (just remove 10-15 lines of code instead of a more complex refactor that needs to directly interact with `IndexInfo`). Apart from `SumSpectraTest` there is only one other unit test that fails (some `Engg*` thing), so that algorithm might be the only one that needs fixing.